src/include/storage/block_compactor.h

```diff
@@ -0,0 +1,99 @@
+#pragma once
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include "storage/arrow_block_metadata.h"
+#include "storage/data_table.h"
+#include "storage/storage_defs.h"
+#include "transaction/transaction_manager.h"
+namespace terrier::storage {
+
+/**
+ * Typedef for a standard hash map with varlen entry as the key. The map uses deep equality checks (whether
+ * the stored underlying byte string is the same) for key comparison.
+ */
+template <class T>
+using VarlenEntryMap = std::unordered_map<VarlenEntry, T, VarlenContentHasher, VarlenContentDeepEqual>;
+
+/**
+ * The block compactor is responsible for taking hot data blocks that are considered to be cold, and make them
+ * arrow-compatible. In the process, any gaps resulting from deletes or aborted transactions are also eliminated.
+ * If the compaction is successful, the block is considered to be fully cold and will be accessed mostly as read-only
+ * data.
+ */
+class BlockCompactor {
+ private:
+  // A Compaction group is a series of blocks all belonging to the same data table. We compact them together
+  // so slots can be freed up. If we only eliminate gaps, deleted slots will never be reclaimed.
+  struct CompactionGroup {
+    CompactionGroup(transaction::TransactionContext *txn, DataTable *table)
+        : txn_(txn),
+          table_(table),
+          all_cols_initializer_(ProjectedRowInitializer::Create(table_->accessor_.GetBlockLayout(),
+                                                                table_->accessor_.GetBlockLayout().AllColumns())),
+          read_buffer_(all_cols_initializer_.InitializeRow(
+              common::AllocationUtil::AllocateAligned(all_cols_initializer_.ProjectedRowSize()))) {}
+
+    ~CompactionGroup() {
+      // Deleting nullptr is just a noop
```

Comment:
I think we tested it and it will cause an ASAN warning though. IIRC we can't really win in this codebase. tidy says nullptr check is unnecessary before delete, while ASAN will fail if you delete nullptr.